tools/nixpkgs-hammer

```diff
@@ -1,20 +1,94 @@
 #!/usr/bin/env python3
 
-import os
-from pathlib import Path
 import argparse
+import io
 import json
+import os
 import subprocess
 import sys
 import textwrap
-from typing import Dict, List, Optional
 from collections import defaultdict
+from dataclasses import asdict, dataclass, field, is_dataclass
+from pathlib import Path
+from typing import Any, Dict, Iterable, List, Optional, Set, Tuple
+
+
+@dataclass(frozen=True)
+class SourceLocation:
+    file: Path
+    line: int
+    column: Optional[int] = field(default=None)
+
+    def stringify(self) -> str:
```

Comment:
Maybe something like render_code_fragment would express that this is not just `__str__` returning `${file}:${column}`.

Also I would probably make this a plain data object, leaving the responsibility of presentation  to a separate function instead of a method.

Reply:
The advantage of having it as a method is that I can give the method the same name on the `SourceLocation` class and on the `Report` class to emphasize their parallelism.

Reply:
I do not particularly see the reason to emphasize that both functions are used to compose the final human-readable report. Either way, the `render_code_fragment` name does not make much sense for the other `stringify` method – I would call it something like `render_report_for_human_terminal` or `render_report_ansi`.

My opposition probably stems from single responsibility principle – it just feels weird having the data containers also handle (just) one method of their rendering. But that is just a feeling so not that important.

Reply:
Okay, I'll move them out of the class into free functions.

Reply:
Done